README.md
# NAME

App::Installdeps - a wrapper of carton for git/dist URL syntax

# SYNOPSIS

    > cat cpanfile
    requires 'Plack';
    requires 'My::Module', git => 'git://github.com/you/My-Module.git', ref => 'develop';
    requires 'My::WebApp', dist => 'http://example.com/My-WebApp.tar.gz';

    > installdeps.pl
    Successfully installed My-Module-0.01
    Successfully installed My-WebApp-1.001
    2 distributions installed
    Installing modules using /home/you/cpanfile
    ...
    Successfully installed Plack-1.0037
    Complete! Modules were installed into /home/you/local

# DESCRIPTION

App::Installdeps is a wrapper of [Carton](https://metacpan.org/pod/Carton) for `git`/`dist` URL syntax.
Unfortunately Carton does not support `git`/`dist` URL syntax yet.
I'm sure it supports them soon!

Meanwhile, why don't you use `installdeps.pl`
which just installs modules with `git`/`dist` URL syntax before executing `carton install`.

# AUTHOR

Shoichi Kaji <skaji@cpan.org>

# COPYRIGHT

Copyright 2015- Shoichi Kaji

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO

[Carton](https://metacpan.org/pod/Carton)


Changes
Revision history for App-Installdeps

{{$NEXT}}
        - Initial release


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.037.
use strict;
use warnings;

use 5.008005;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "a wrapper of carton for git/dist URL syntax",
  "AUTHOR" => "Shoichi Kaji <skaji\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "App-Installdeps",
  "EXE_FILES" => [
    "script/installdeps.pl"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008005",
  "NAME" => "App::Installdeps",
  "PREREQ_PM" => {
    "App::cpanminus" => 0,
    "Carton::CLI" => 0,
    "Module::CPANfile" => 0,
    "Module::Metadata" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => "0.96"
  },
  "VERSION" => "0.01",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "App::cpanminus" => 0,
  "Carton::CLI" => 0,
  "ExtUtils::MakeMaker" => 0,
  "Module::CPANfile" => 0,
  "Module::Metadata" => 0,
  "Test::More" => "0.96"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


script/installdeps.pl
#!perl
use strict;
use warnings;
use utf8;
use App::cpanminus::fatscript;
use Carton::CLI;
use Config;
use Getopt::Long ();
use Module::CPANfile;
use Module::Metadata;
use Pod::Usage ();

my $p = Getopt::Long::Parser->new(
    config => [ "no_ignore_case", "no_auto_abbrev" ],
);

$p->getoptionsfromarray(
    \@ARGV,
    "h|help"    => sub { Pod::Usage::pod2usage(0) },
    "reinstall" => \(my $reinstall),
) or exit 1;

my $cpanfile = Module::CPANfile->load("cpanfile");
my @module = $cpanfile->merged_requirements->required_modules;
my $inc = [grep -d, "local/lib/perl5", "local/lib/perl5/$Config{archname}"];

my @non_cpan;
for my $module (sort @module) {
    my $prereq = $cpanfile->prereq_for_module($module);
    my $options = $prereq->requirement->options;
    if (my $dist = $options->{dist}) {
        if (!$reinstall) {
            my $meta = Module::Metadata->new_from_module($module, inc => $inc);
            next if $meta;
        }
        push @non_cpan, $dist;
    } elsif (my $git = $options->{git}) {
        if ($git =~ /^http/i) {
            die "${module}'s git url must start with git://, but '$git'\n";
        }
        if (!$reinstall) {
            my $meta = Module::Metadata->new_from_module($module, inc => $inc);
            next if $meta;
        }
        if (my $ref = $options->{ref}) {
            $git .= "\@$ref";
        }
        push @non_cpan, $git;
    }
}

if (@non_cpan) {
    my $cpanm = Module::Metadata->find_module_by_name("App::cpanminus::fatscript");
    !system $^X, $cpanm, "-qn", "-Llocal", @non_cpan or exit 1;
}
exit Carton::CLI->new->run("install");

__END__

=head1 NAME

installdeps.pl - a wrapper of carton for git/dist URL syntax

=head1 SYNOPSIS

  > installdeps.pl [--reinstall]

  Options:
  -h, --help       Show this help
      --reinstall  Reinstall modules even if they are already installed

=head1 AUTHOR

Shoichi Kaji E<lt>skaji@cpan.orgE<gt>

=head1 COPYRIGHT

Copyright 2015- Shoichi Kaji

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/App/Installdeps.pm
package App::Installdeps;

use strict;
use 5.008_005;
our $VERSION = '0.01';

1;
__END__

=encoding utf-8

=head1 NAME

App::Installdeps - a wrapper of carton for git/dist URL syntax

=head1 SYNOPSIS

  > cat cpanfile
  requires 'Plack';
  requires 'My::Module', git => 'git://github.com/you/My-Module.git', ref => 'develop';
  requires 'My::WebApp', dist => 'http://example.com/My-WebApp.tar.gz';

  > installdeps.pl
  Successfully installed My-Module-0.01
  Successfully installed My-WebApp-1.001
  2 distributions installed
  Installing modules using /home/you/cpanfile
  ...
  Successfully installed Plack-1.0037
  Complete! Modules were installed into /home/you/local

=head1 DESCRIPTION

App::Installdeps is a wrapper of L<Carton> for C<git>/C<dist> URL syntax.
Unfortunately Carton does not support C<git>/C<dist> URL syntax yet.
I'm sure it supports them soon!

Meanwhile, why don't you use C<installdeps.pl>
which just installs modules with C<git>/C<dist> URL syntax before executing C<carton install>.

=head1 AUTHOR

Shoichi Kaji E<lt>skaji@cpan.orgE<gt>

=head1 COPYRIGHT

Copyright 2015- Shoichi Kaji

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<Carton>

=cut
